pommerman/forward_model.py

```diff
@@ -156,128 +156,256 @@ def step(actions,
+                    crossings[border] = (bomb_num, False)
+
+        # Deal with multiple agents or multible bomb collisions on desired next position
+        # by resetting desired position to current position for everyone involved in the collision
+        agent_occupancy = np.zeros_like(curr_board)
+        bomb_occupancy = np.zeros_like(curr_board)
+
+        for desired_position in desired_agent_positions:
+            agent_occupancy[desired_position] += 1
+
+        for desired_position in desired_bomb_positions:
+            bomb_occupancy[position] += 1
+
+        # Resolve two or more agents or two or more bombs trying
+        # to occupy the same space.
+        # We will deal with an agent and a bomb trying to share space later
+        change = True
+        while change:
+            change = False
+            for agent_num, agent in enumerate(alive_agents):
+                desired_position = desired_agent_positions[agent_num]
+                curr_position = agent.position
+                if desired_position != curr_position and \
+                      (agent_occupancy[desired_position] > 1 or bomb_occupancy[desired_position] > 1):
+                    desired_agent_positions[agent_num] = curr_position
+                    agent_occupancy[curr_position] += 1
+                    change = True
+
+            for bomb_num, bomb in enumerate(curr_bombs):
+                desired_position = desired_bomb_positions[bomb_num]
+                curr_position = bomb.position
+                if desired_position != curr_position and \
+                      (bomb_occupancy[desired_position] > 1 or agent_occupancy[desired_position] > 1):
+                    desired_bomb_positions[bomb_num] = curr_position
+                    bomb_occupancy[curr_position] += 1
+                    change = True
+
+        # Deal with kicks
+
+        bombs_kicked_by = dict()
+        delayed_bomb_updates = []
+        delayed_agent_updates = []
+
+        # Loop through all bombs to see if they need a good kicking or cause collisions with an agent
+        for bomb_num, bomb in enumerate(curr_bombs):
+            desired_position = desired_bomb_positions[bomb_num]
+
+            if agent_occupancy[desired_position] == 0:
+                # There was never an agent around to kick us or to collide with us
+                continue
 
-        for num_agent, agent in enumerate(curr_agents):
-            if not agent.is_alive:
+            agent_list = [(agent_num, agent) for (agent_num, agent) in enumerate(alive_agents) \
+                         if desired_position == desired_agent_positions[agent_num]]
+            if not agent_list:
+                # Agents moved from collision
                 continue
 
-            for num_agent2 in range(num_agent + 1, len(curr_agents)):
-                agent2 = curr_agents[num_agent2]
-                if not agent2.is_alive:
+            # Should contain a single element
+            assert len(agent_list) == 1
```

Comment:
when does this not happen?

Reply:
(referring to the assert)

Reply:
In the kick phase we should have at most a single agent sharing a square with a ball.
If we have more then there is a bug :-)